src/core/device.c

```diff
@@ -90,6 +92,46 @@ static int device_set_sysfs(Device *d, const char *sysfs) {
         return 0;
 }
 
+static void io_cost_qos_reset(IOCostQos *io_cost_qos) {
+        assert(io_cost_qos);
+
+        io_cost_qos->enabled = -1;
+        io_cost_qos->ctrl = _IO_COST_CTRL_INVALID;
+        io_cost_qos->read_latency_percentile = 0;
+        io_cost_qos->read_latency_threshold = 0;
+        io_cost_qos->write_latency_percentile= 0;
+        io_cost_qos->write_latency_threshold= 0;
+        io_cost_qos->min= 0;
+        io_cost_qos->max= 0;
+        /* If either min or max is set to a value below 1
+         * the kernel will set to 1 */
+}
+
+static void io_cost_model_reset(IOCostModel *io_cost_model) {
+        assert(io_cost_model);
+
+        io_cost_model->ctrl = _IO_COST_CTRL_INVALID;
+        io_cost_model->rbps = 0;
+        io_cost_model->rseqiops = 0;
+        io_cost_model->rrandiops = 0;
+        io_cost_model->wbps = 0;
+        io_cost_model->wseqiops = 0;
+        io_cost_model->wrandiops = 0;
+}
+
+static bool verify_io_cost_supported(Device *d) {
+        /* io_cost_qos and io_cost_model only apply to cgroup v2 */
+        if (!cg_all_unified())
+                return false;
+
+        if (!d || !d->devname) {
```

Comment:
We debug log if `d` is `NULL` here but at the call site just before this function is called we have `assert(d)`. So shouldn't we just have `assert(d)` in this function as well?

Reply:
The primary purpose is to check devname, which can still be null. I do not believe we would want to simply repeat `assert(d)`.